linkerd/timeout/src/failfast.rs

```diff
@@ -0,0 +1,191 @@
+
+                    // A timeout has been set, so wait for it to complete.
+                    State::Waiting(ref mut fut) => {
+                        let poll = fut.poll();
+                        debug_assert!(poll.is_ok(), "timer must not fail");
+                        if let Ok(Async::NotReady) = poll {
+                            return Ok(Async::NotReady);
+                        }
+                        State::FailFast
+                    }
+
+                    // Admit requests and fail them immediately.
+                    State::FailFast => {
+                        return Ok(Async::Ready(()));
+                    }
+                };
+            },
+
+            // If the inner service is ready or has failed, then let subsequent
+            // calls through to the service.
+            ret => {
+                self.state = State::Open;
+                return ret.map_err(Into::into);
+            }
+        }
+    }
+
+    fn call(&mut self, req: T) -> Self::Future {
+        match self.state {
+            State::Open => ResponseFuture::Inner(self.inner.call(req)),
+            State::FailFast => ResponseFuture::FailFast,
+            State::Waiting(_) => panic!("poll_ready must be called"),
+        }
+    }
+}
+
+impl<F> Future for ResponseFuture<F>
+where
+    F: Future,
+    F::Error: Into<Error>,
+{
+    type Item = F::Item;
+    type Error = Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        match self {
+            ResponseFuture::Inner(ref mut f) => f.poll().map_err(Into::into),
+            ResponseFuture::FailFast => Err(FailFastError(()).into()),
+        }
+    }
+}
+
+// === impl FailFastError ===
+
+impl std::fmt::Display for FailFastError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Service in fail-fast")
+    }
+}
+
+impl std::error::Error for FailFastError {}
+
+#[cfg(test)]
+mod test {
+    use super::FailFastLayer;
+    use futures::{future, Future};
+    use std::time::{Duration, Instant};
+    use tower::layer::Layer;
+    use tower::Service;
+    use tower_test::mock;
+
+    fn run<F, R>(f: F)
+    where
+        F: FnOnce() -> R + 'static,
+        R: future::IntoFuture<Item = ()> + 'static,
+    {
+        tokio::runtime::current_thread::run(future::lazy(f).map_err(|_| panic!("Failed")));
+    }
+
+    #[test]
+    fn fails_fast() {
+        let max_unavailable = Duration::from_millis(100);
+        let (service, mut handle) = mock::pair::<(), ()>();
+        let mut service = FailFastLayer::new(max_unavailable).layer(service);
+
+        run(move || {
+            // The inner starts unavailable.
+            handle.allow(0);
+            assert!(service.poll_ready().unwrap().is_not_ready());
+
+            // Then we wait for the idle timeout, at which point the service
+            // should start failing fast.
+            tokio_timer::Delay::new(Instant::now() + max_unavailable + Duration::from_millis(1))
```

Comment:
Do you have any examples of this? I'm open to it but these tests seem Good Enough, since we have a single thread and aren't really exposing ourselves to anything overtly racey